novaan-server/src/Auth/Jwt/JwtService.cs

```diff
@@ -0,0 +1,200 @@
+                    return new SignInResponseDTO()
+                    {
+                        Result = false,
+                        Errors = new List<string>()
+                        {
+                            "Invalid tokens"
+                        }
+                    };
+
+                if (storedToken.IsUsed)
+                    return new SignInResponseDTO()
+                    {
+                        Result = false,
+                        Errors = new List<string>()
+                        {
+                            "Invalid tokens"
+                        }
+                    };
+
+                if (storedToken.IsRevoked)
+                    return new SignInResponseDTO()
+                    {
+                        Result = false,
+                        Errors = new List<string>()
+                        {
+                            "Invalid tokens"
+                        }
+                    };
+
+                var jti = tokenInVerification.Claims.FirstOrDefault(x => x.Type == JwtRegisteredClaimNames.Jti).Value;
+                if (storedToken.JwtId != jti)
+                    return new SignInResponseDTO()
+                    {
+                        Result = false,
+                        Errors = new List<string>()
+                        {
+                            "Invalid tokens"
+                        }
+                    };
+
+                if (storedToken.ExpiryDate < DateTime.UtcNow)
+                    return new SignInResponseDTO()
+                    {
+                        Result = false,
+                        Errors = new List<string>()
+                        {
+                            "Expired tokens"
+                        }
+                    };
+
+                var filterStoredToken = Builders<RefreshToken>.Filter.Eq("Token", tokenRequest.RefreshToken);
+                var updateStoredToken = Builders<RefreshToken>.Update.Set("IsUsed", true);
+                await _mongoDBService.RefreshTokens.UpdateOneAsync(filterStoredToken, updateStoredToken);
+
+
+                return await GennerateJwtToken(new UserJwt()
+                {
+                    Email = storedToken.UserMail,
+                });
+
+            }
+            catch (Exception e)
+            {
+                return new SignInResponseDTO()
+                {
+                    Result = false,
+                    Errors = new List<string>()
+                        {
+                            "Server error"
+                        }
+                };
+            }
+        }
+
+        private DateTime unixTimeStampToDateTime(long unixTimeStamp)
+        {
+            var dateTimeVal = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
+            dateTimeVal = dateTimeVal.AddSeconds(unixTimeStamp).ToUniversalTime();
+
+            return dateTimeVal;
+        }
+
+        private string randomStringGeneration(int length)
```

Comment:
Đang cảm giác hàm nãy sẽ khá tốn tài nguyên xử lí. Suggest chuyển sang hash some unique thing để sử dụng làm RefreshToken

Reply:
mục đích a đang muốn tạo ra một chuỗi random, nếu muốn vẫn có thể băm chuỗi random vừa tạo với thuật toán SHA256